### PR TITLE
refactor(react-sdk): enumerate devices when components that need them are rendered

### PR DESCRIPTION
**Pull request by MartinCupela**
## Refactor
1. Move the device enumeration effects into hooks that are invoked in components that actually need access to  up-to-date array of devices of a specific kind. Remove enumerate prop from `MediaDeviceContextProps`.
2. Rename `MediaDeviceContext` API methods that toggle initial device state to reflect they are touching only initial device state.

## Bug fix

1. When publishing audio / video as a result of click on toggle buttons, set the initial video/audio state to enabled. This will then allow republishing streams, when selected device id changes. 
2. Allow ParticipantView to shrink if the container is smaller

### CODE REVIEW

File: packages/styling/src/ParticipantView/ParticipantView-layout.scss [hunk 1]
```diff
@@ -6,6 +6,7 @@ $scope-name: 'str-video__participant-view';
   width: 100%;
   max-width: 1280px;
   max-height: 100%;
+  min-height: 0;
```
Comment (arnautov-anton) [inline packages/styling/src/ParticipantView/ParticipantView-layout.scss:9]: Is this a change for a specific layout?

Comment (MartinCupela) [reply packages/styling/src/ParticipantView/ParticipantView-layout.scss:9]: This is reaction to the component overflow in messenger clone.

File: packages/react-sdk/src/core/hooks/useObserveDevices.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import {
+  getAudioDevices,
+  getAudioOutputDevices,
+  getVideoDevices,
+} from '@stream-io/video-client';
+
+/**
+ * Observes changes in connected devices and maintains an up-to-date array of connected MediaDeviceInfo objects.
+ * @param observeDevices
+ */
```
Comment (szuperaz) [inline packages/react-sdk/src/core/hooks/useObserveDevices.ts:12]: I think the `@category Device Management` lines are missing, so the new hooks are under "Device Management" category in the docs: https://staging.getstream.io/video/docs/react/call-engine/hooks-and-contexts/#device-management

File: packages/react-sdk/src/core/contexts/MediaDevicesContext.tsx [hunk 3]
```diff
@@ -95,13 +97,14 @@ export type MediaDevicesContextAPI = {
...
+  setInitialAudioEnabled: (enabled: boolean) => void;
   setInitialVideoState: (state: DeviceState) => void;
   selectedAudioInputDeviceId?: string;
   selectedAudioOutputDeviceId?: string;
   selectedVideoDeviceId?: string;
   switchDevice: (kind: MediaDeviceKind, deviceId?: string) => void;
-  toggleAudioMuteState: () => void;
-  toggleVideoMuteState: () => void;
+  toggleInitialAudioMuteState: () => void;
+  toggleInitialVideoMuteState: () => void;
```
Comment (oliverlaz) [inline packages/react-sdk/src/core/contexts/MediaDevicesContext.tsx:107]: These are public api methods. Why do they have an `initial` in their name when they are used for regular toggling?

File: packages/styling/src/ParticipantView/ParticipantView-layout.scss (hunk 1 above)
Comment (MartinCupela) [reply packages/styling/src/ParticipantView/ParticipantView-layout.scss:9]: I have removed the rule in CallParticipantsView as you requested :)

File: packages/react-sdk/src/core/contexts/MediaDevicesContext.tsx (hunk 3 above)
Comment (oliverlaz) [reply packages/react-sdk/src/core/contexts/MediaDevicesContext.tsx:107]: Ok, I missed that part. I keep forgetting that both buttons (preview and in-call) live in the same file in the SDK.

**Pull request merged**